From a9f707e86c97781a5a36794f8fb43e061552d946 Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <josh@wilsdon.ca>
Date: Mon, 5 Mar 2018 14:52:10 -0800
Subject: [PATCH] TRITON-181 would like packages to be able to optionally be
 brand-specific Reviewed by: Trent Mick <trentm@gmail.com> Approved by: Trent
 Mick <trentm@gmail.com>

---
 docs/index.md                | 105 +++++++++++++++++++----------------
 lib/validations.js           |   7 ++-
 package.json                 |   2 +-
 sapi_manifests/papi/template |   6 +-
 test/api.test.js             |  32 +++++++++--
 tools/rsync-to               |  66 ++++++++++++++++++++++
 6 files changed, 162 insertions(+), 56 deletions(-)
 create mode 100755 tools/rsync-to

diff --git a/docs/index.md b/docs/index.md
index 0df9a81..1ff0788 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -1,5 +1,5 @@
 ---
-title: SDC 7 Package API
+title: Triton Package API
 markdown2extras: tables, code-friendly
 apisections: PackageObjects, Packages, Ping, Changelog
 ---
@@ -10,24 +10,21 @@ apisections: PackageObjects, Packages, Ping, Changelog
 -->
 
 <!--
-    Copyright (c) 2016, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
-# SDC 7 Package API
+# Triton Package API
 
-The specifications used by SDC internally to create a machine from a given image
-are called packages, and are referred to as packages by all the APIs. For
-example, packages can specify the amount of RAM and CPU a new machine will use.
-The word 'packages' is independent of the name chosen by marketing for the
-outside world.
+The collections of properties used by Triton internally to create a VM are
+called "packages", and are referred to as packages by all the other APIs. For
+example, packages can specify the amount of RAM and CPU a new machine will use,
+and what the disk quota will be.
 
 Some of the package attributes are used by `vmadm` to create or resize machines.
 Please refer to the [vmadm man page](https://github.com/joyent/smartos-live/blob/master/src/vm/man/vmadm.1m.md#properties)
 to review the meaning of these properties for the machines.
 
 
-
-
 # Package objects
 
 Package entries are stored in Moray in the form of JSON objects. Here is an
@@ -35,55 +32,55 @@ example of a package:
 
 
     {
-        v: 1,
-        uuid: "7fc87f43-2def-4e6f-9f8c-980b0385b36e",
         active: true,
         cpu_cap: 25,
-        group: "Standard",
+        cpu_shares: 25
         description: "Micro 0.25 GB RAM 0.125 CPUs 16 GB Disk",
+        group: "Standard",
         max_lwps: 4000,
         max_physical_memory: 256,
         max_swap: 512,
         name: "g3-standard-0.25-smartos",
-        common_name: "Standard 0.25",
-        quota: 16384,
         networks: ["1e7bb0e1-25a9-43b6-bb19-f79ae9540b39", "193d6804-256c-4e89-a4cd-46f045959993"],
+        quota: 16384,
+        uuid: "7fc87f43-2def-4e6f-9f8c-980b0385b36e",
+        v: 1,
         version: "1.0.0",
-        zfs_io_priority: 100,
-        fss: 25,
-        cpu_burst_ratio: 0.5,
-        ram_ratio: 1.995012469
+        zfs_io_priority: 100
     }
 
 
-| Attribute                                           | Required  | Unique | Immutable | Type    | Explanation                                                                                                |
-| --------------------------------------------------- | --------- | ------ | --------- | ------- | ---------------------------------------------------------------------------------------------------------- |
-| [active](#package-active)                           | true      |        |           | boolean | Whether it can currently be used for provisioning.                                                         |
-| [billing_tag](#package-billing_tag)                 |           |        |           | string  | Arbitrary tag that can be used by ops for billing purposes; it has no intrinsic meaning to SDC.            |
-| [common_name](#package-common_name)                 |           |        |           | string  | Name displayed in the Portal.                                                                              |
-| cpu_burst_ratio                                     |           |        |           | float   | Typically computed value. See below for more.                                                              |
-| [cpu_cap](#package-cpu_cap)                         | sometimes |        | true      | integer | Cap on how much CPU a machine can use. 100 = one core, 350 = 3.5 cores, etc.                               |
-| [default](#package-default)                         |           |        |           | boolean | **DEPRECATED** Whether this is the default package of this name through the SDC 6.5 API                    |
-| [description](#package-description)                 |           |        |           | string  | Human description of this package.                                                                         |
-| [fss](#package-fss)                                 |           |        |           | integer | CPU shares for a VM. This operates relative to other machines on a CN. (also known as cpu_shares)          |
-| [group](#package-group)                             |           |        |           | string  | Group of associated packages. E.g. High CPU, High Memory, High Storage, High IO or the customer's name.    |
-| [max_lwps](#package-max_lwps)                       | true      |        | true      | integer | Max number of processes allowed                                                                            |
-| [max_physical_memory](#package-max_physical_memory) | true      |        | true      | integer | Max RAM in MiB.                                                                                            |
-| [max_swap](#package-max_swap)                       | true      |        | true      | integer | Max swap in MiB.                                                                                           |
-| [min_platform](#package-min_platform)               |           |        |           | hash    | Minimum version(s) of OS platforms that this package can use.                                              |
-| [name](#package-name)                               | true      |        | true      | string  | Name of package in API. See below for details on valid names.                                              |
-| [networks](#package-networks)                       |           |        |           | array   | UUIDs of networks that the machine requires access to.                                                     |
-| [os](#package-os)                                   |           |        | true      | string  | Operating system for this package.                                                                         |
-| [owner_uuids](#package-owner_uuids)                 |           |        |           | array   | UUIDs of package owners.                                                                                   |
-| [parent](#package-parent)                           |           |        |           | string  | `name` of instance this was cloned from. Useful if package is created from another package for a customer. |
-| [quota](#package-quota)                             | true      |        | true      | integer | Disk size in MiB. Must be a multiple of 1024.                                                              |
-| ram_ratio                                           |           |        |           | float   | Typically computed value. See below for more.                                                              |
-| [traits](#package-traits)                           |           |        |           | hash    | Set of traits for provisioning to servers. See DAPI docs for details on traits.                            |
-| [uuid](#package-uuid)                               | true      | true   | true      | uuid    | Package identifier.                                                                                        |
-| [v](#package-v)                                     |           |        |           | integer | API version of PAPI.                                                                                       |
-| [version](#package-version)                         | true      |        | true      | string  | Semver version number.                                                                                     |
-| [vcpus](#package-vcpus)                             | sometimes |        | true      | integer | Number of cpus to show, between 1 - 64. Required during provisioning if `type` == 'kvm'.                   |
-| [zfs_io_priority](#package-zfs_io_priority)         | true      |        | true      | integer | ZFS I/O priority. This operates relative to other machines on a CN, determining which get I/O first.       |
+## Attributes
+
+| Attribute                                           | Required  | Unique | Immutable | Type    | Added In | Explanation                                                                                                |
+| --------------------------------------------------- | --------- | ------ | --------- | ------- | -------- | ---------------------------------------------------------------------------------------------------------- |
+| [active](#package-active)                           | true      |        |           | boolean |          | Whether it can currently be used for provisioning.                                                         |
+| [billing_tag](#package-billing_tag)                 |           |        |           | string  |          | Arbitrary tag that can be used by ops for billing purposes; it has no intrinsic meaning to Triton.         |
+| [brand](#package-brand)                             |           |        | true      | string  |   v7.1.0 | Force this brand for zones using this package, one of: 'bhyve', 'joyent', 'joyent-minimal', 'kvm', 'lx'    |
+| [common_name](#package-common_name)                 |           |        |           | string  |          | Name displayed in the Portal.                                                                              |
+| cpu_burst_ratio                                     |           |        |           | float   |          | Typically computed value. See below for more.                                                              |
+| [cpu_cap](#package-cpu_cap)                         | sometimes |        | true      | integer |          | Cap on how much CPU a machine can use. 100 = one core, 350 = 3.5 cores, etc.                               |
+| [default](#package-default)                         |           |        |           | boolean |          | **DEPRECATED** Whether this is the default package of this name through the SDC 6.5 API                    |
+| [description](#package-description)                 |           |        |           | string  |          | Human description of this package.                                                                         |
+| [fss](#package-fss)                                 |           |        |           | integer |          | CPU shares for a VM. This operates relative to other machines on a CN. (also known as cpu_shares)          |
+| [group](#package-group)                             |           |        |           | string  |          | Group of associated packages. E.g. High CPU, High Memory, High Storage, High IO or the customer's name.    |
+| [max_lwps](#package-max_lwps)                       | true      |        | true      | integer |          | Max number of processes allowed                                                                            |
+| [max_physical_memory](#package-max_physical_memory) | true      |        | true      | integer |          | Max RAM in MiB.                                                                                            |
+| [max_swap](#package-max_swap)                       | true      |        | true      | integer |          | Max swap in MiB.                                                                                           |
+| [min_platform](#package-min_platform)               |           |        |           | hash    |          | Minimum version(s) of OS platforms that this package can use.                                              |
+| [name](#package-name)                               | true      |        | true      | string  |          | Name of package in API. See below for details on valid names.                                              |
+| [networks](#package-networks)                       |           |        |           | array   |          | UUIDs of networks that the machine requires access to.                                                     |
+| [os](#package-os)                                   |           |        | true      | string  |          | Operating system for this package.                                                                         |
+| [owner_uuids](#package-owner_uuids)                 |           |        |           | array   |          | UUIDs of package owners.                                                                                   |
+| [parent](#package-parent)                           |           |        |           | string  |          | `name` of instance this was cloned from. Useful if package is created from another package for a customer. |
+| [quota](#package-quota)                             | true      |        | true      | integer |          | Disk size in MiB. Must be a multiple of 1024.                                                              |
+| ram_ratio                                           |           |        |           | float   |          | Typically computed value. See below for more.                                                              |
+| [traits](#package-traits)                           |           |        |           | hash    |          | Set of traits for provisioning to servers. See DAPI docs for details on traits.                            |
+| [uuid](#package-uuid)                               | true      | true   | true      | uuid    |          | Package identifier.                                                                                        |
+| [v](#package-v)                                     |           |        |           | integer |          | API version of PAPI.                                                                                       |
+| [version](#package-version)                         | true      |        | true      | string  |          | Semver version number.                                                                                     |
+| [vcpus](#package-vcpus)                             | sometimes |        | true      | integer |          | Number of cpus to show, between 1 - 64. Required during provisioning if `type` == 'kvm'.                   |
+| [zfs_io_priority](#package-zfs_io_priority)         | true      |        | true      | integer |          | ZFS I/O priority. This operates relative to other machines on a CN, determining which get I/O first.       |
 
 
 ## Package: active
@@ -98,6 +95,20 @@ If true, this package can be used for provisioning, otherwise not.
 An arbitrary string that can be used by operators for billing purposes. This is
 an opaque string, where no special meaning is enforced by SDC.
 
+## Package: brand
+
+Added In: v7.1.0
+
+This optional parameter ties a package to a specific zone brand. The brand is
+what's used to determine which type of virtualization to use for the instance.
+Most commonly this should be set to one of: 'bhyve' or 'kvm' when a datacenter
+supports both types of virtualization and a package is being used for one or the
+other.
+
+The value must be one of: 'bhyve', 'joyent', 'joyent-minimal', 'kvm', 'lx'.
+
+    "brand": "kvm"
+
 
 ## Package: common_name
 
diff --git a/lib/validations.js b/lib/validations.js
index 0683619..0f583b0 100644
--- a/lib/validations.js
+++ b/lib/validations.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -18,6 +18,7 @@ var sprintf = util.format;
 var BAD_NAME_RE = /[\_\-\.][\_\-\.]/;
 var NAME_RE = /^[a-zA-Z0-9]([a-zA-Z0-9\_\-\.]+)?[a-zA-Z0-9]$/;
 var UUID_RE = /^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/;
+var VALID_BRANDS = ['bhyve', 'joyent', 'joyent-minimal', 'kvm', 'lx'];
 var VALID_SPREADS = ['min-ram', 'random', 'min-owner'];
 
 ///--- Validation helpers
@@ -170,6 +171,10 @@ function validate(pkg, schema) {
                 VALID_SPREADS.join(', '));
     }
 
+    if (pkg.brand !== undefined && VALID_BRANDS.indexOf(pkg.brand) === -1) {
+        invalid('brand', 'must be one of: ' + VALID_BRANDS.join(', '));
+    }
+
     if (!validNumber(pkg.max_physical_memory, MIN_RAM)) {
         invalid('max_physical_memory',
                 'must be greater or equal to ' + MIN_RAM);
diff --git a/package.json b/package.json
index b836806..250b4dc 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-papi",
   "description": "Packages API",
-  "version": "7.0.0",
+  "version": "7.1.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "repository": {
diff --git a/sapi_manifests/papi/template b/sapi_manifests/papi/template
index 87d8315..9e26247 100644
--- a/sapi_manifests/papi/template
+++ b/sapi_manifests/papi/template
@@ -6,7 +6,7 @@
         "cueballOptions": {
             "resolvers": [ "{{{BINDER_SERVICE}}}" ]
         },
-        "version": 6
+        "version": 7
     },
     "ufds_admin_uuid": "{{{ufds_admin_uuid}}}",
     "first_boot": {{{SERVICE_IS_FIRST_BOOT}}},
@@ -26,6 +26,10 @@
     },
     "bucket": "sdc_packages",
     "schema": {
+        "brand": {
+            "index": true,
+            "type": "string"
+        },
         "uuid": {
             "type": "uuid",
             "unique": true,
diff --git a/test/api.test.js b/test/api.test.js
index 07b6f13..0264569 100644
--- a/test/api.test.js
+++ b/test/api.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -17,12 +17,13 @@ var fs      = require('fs');
 var path    = require('path');
 var qs      = require('querystring');
 var os      = require('os');
-var restify = require('restify');
-var test    = require('tape').test;
 var util    = require('util');
-var Logger  = require('bunyan');
-var libuuid = require('libuuid');
+
 var jsprim = require('jsprim');
+var libuuid = require('libuuid');
+var Logger  = require('bunyan');
+var restify = require('restify');
+var test    = require('tape').test;
 var VError = require('verror');
 
 var papi = require('../lib/papi');
@@ -32,7 +33,7 @@ var papi = require('../lib/papi');
 // someone's laptop and they're using the default COAL IP addresses.
 var cfgFile = path.resolve(__dirname, '../etc/config.json');
 cfgFile = fs.existsSync(cfgFile) ? cfgFile :
-               path.resolve(__dirname, '../etc/config.test.json');
+    path.resolve(__dirname, '../etc/config.test.json');
 
 var config = JSON.parse(fs.readFileSync(cfgFile, 'utf-8'));
 
@@ -105,6 +106,20 @@ var packages = [ {
         '7f5501af-12da-4727-8579-625e527ed1f2',
         'c39b6d6a-1c11-4d1b-b213-174974d71b45'
     ]
+}, {
+    v: 1,
+    active: true,
+    brand: 'bhyve',
+    cpu_cap: 300,
+    max_lwps: 1000,
+    max_physical_memory: 512,
+    max_swap: 1024,
+    name: pkgName('bhyve-512'),
+    quota: 10240,
+    version: '1.0.0',
+    uuid: 'b87479f8-1ce1-11e8-8444-636ba22202b2',
+    vcpus: 2,
+    zfs_io_priority: 100
 } ];
 
 
@@ -369,6 +384,7 @@ test('POST /packages (empty required fields)', function (t) {
 
 test('POST /packages (fields validation failed)', function (t) {
     var pkg = {
+        brand: 'WcDonalds',
         name: pkgName('fail-validation'),
         version: '1.0.0',
         os: 2,
@@ -414,6 +430,10 @@ test('POST /packages (fields validation failed)', function (t) {
             { field: 'alloc_server_spread',
               code: 'Invalid',
               message: 'must be one of: min-ram, random, min-owner' },
+            { field: 'brand',
+              code: 'Invalid',
+              message:
+                'must be one of: bhyve, joyent, joyent-minimal, kvm, lx' },
             { field: 'max_physical_memory',
               code: 'Invalid',
               message: 'must be greater or equal to 64' },
diff --git a/tools/rsync-to b/tools/rsync-to
new file mode 100755
index 0000000..621f4c4
--- /dev/null
+++ b/tools/rsync-to
@@ -0,0 +1,66 @@
+#!/bin/bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright (c) 2018, Joyent, Inc.
+#
+
+#
+# Rsync the master in this working copy to the install on the given HN.
+#
+
+if [[ -n "$TRACE" ]]; then
+    export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+    set -o xtrace
+fi
+set -o errexit
+set -o pipefail
+
+
+TOP=$(cd $(dirname $0)/../; pwd)
+NODE="root@$1"
+
+if [[ -z "$PAPI_ZONE" ]]; then
+    PAPI_ZONE=$(ssh $NODE "vmadm lookup -1 alias=~^papi" 2>/dev/null)
+fi
+echo "PAPI_ZONE: $PAPI_ZONE"
+
+extraOpts=
+if [[ $(uname -s) != "SunOS" ]]; then
+    extraOpts="--exclude *.node --exclude build"
+else
+    # Clean node_modules everytime.
+    ssh $NODE rm -rf /zones/$PAPI_ZONE/root/opt/smartdc/papi/node_modules
+fi
+
+rsync -av ${TOP}/ \
+    $NODE:/zones/$PAPI_ZONE/root/opt/smartdc/papi/ \
+    $extraOpts \
+    --exclude .git/ \
+    --exclude .gitignore \
+    --exclude .gitmodules \
+    --exclude /boot/ \
+    --exclude /deps/ \
+    --exclude /docs/ \
+    --exclude Makefile \
+    --exclude /tools/
+
+rsync -av ${TOP}/boot/ \
+    $NODE:/zones/$PAPI_ZONE/root/opt/smartdc/boot/ \
+    $extraOpts
+
+# disable errexit because grep exits non-zero when it doesn't find the error
+set +o errexit
+
+ssh ${NODE} \
+    "svcadm -z ${PAPI_ZONE} clear papi 2>&1 || svcadm -z ${PAPI_ZONE} restart papi" \
+    | grep -v "is not in a maintenance or degraded state."
+
+ssh ${NODE} \
+    "svcadm -z ${PAPI_ZONE} clear config-agent 2>&1 || svcadm -z ${PAPI_ZONE} restart config-agent" \
+    | grep -v "is not in a maintenance or degraded state."
+
-- 
2.21.0

